# Exploring eBay Car Sales Data

#### Table of Contents
* [Introduction](#1)
* [Exploratory Data Analysis](#2)
* [Data Cleaning](#3)
* [Data Analysis](#4)
* [Conclusion](#5)

#### Introduction <a class='anchor' id='1'></a>

In this project, we explore the eBay car sales dataset from DataQuest guided project.

The objective of this project:
- To practice Exploratory Data Analysis
- To practice Data Cleaning

#### Exploratory Data Analysis <a class='anchor' id='2'></a>

In [195]:
import pandas as pd, numpy as numpy

autos = pd.read_csv("D:/DataQuest/autos.csv", encoding='Latin-1')

File couldn't be read with UTF-8. Successful with latin-1

In [49]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

50,000 total rows
vehicleType, gearbox, model, fuleType, notRepairedDamage - these may have null values or missing data
dateCrawled, dateCreated, lastSeen - these may need to be converted to datetime
price - need to remove $ and converted to float
odometer - need to remove km and comma

column names can be renamed

#### Data Cleaning <a class='anchor' id='3'></a>

#### rename columns

We will change camelcase names to snakecase names.

In [196]:
autos.rename({'yearOfRegistration':'registration_year'}, axis=1, inplace=True)
autos.rename({'monthOfRegistration':'registration_month'}, axis=1, inplace=True)
autos.rename({'notRepairedDamage':'unrepaired_damage'}, axis=1,inplace=True)
autos.rename({'dateCreated':'ad_created'}, axis=1,inplace=True)
autos.rename({'dateCrawled':'date_crawled'}, axis=1,inplace=True)
autos.rename({'offerType':'offer_type'}, axis=1,inplace=True)
autos.rename({'vehicleType':'vehicle_type'}, axis=1,inplace=True)
autos.rename({'powerPS':'power_ps'}, axis=1,inplace=True)
autos.rename({'fuelType':'fuel_type'}, axis=1,inplace=True)
autos.rename({'nrOfPictures':'num_pictures'}, axis=1,inplace=True)
autos.rename({'postalCode':'postal_code'}, axis=1,inplace=True)
autos.rename({'lastSeen':'last_seen'}, axis=1,inplace=True)
autos.head(2)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   

  offer_type   price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot  $5,000  control          bus               2004    manuell   
1    Angebot  $8,500  control    limousine               1997  automatik   

   power_ps   model   odometer  registration_month fuel_type    brand  \
0       158  andere  150,000km                   3       lpg  peugeot   
1       286     7er  150,000km                   6    benzin      bmw   

  unrepaired_damage           ad_created  num_pictures  postal_code  \
0              nein  2016-03-26 00:00:00             0        79588   
1              nein  2016-04-04 00:00:00             0        71034   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08

#### Data exploration continued...

In [36]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

We can possibly remove seller and offer type, as there are only 2 unique types and the top frequency is 49999 (basiclaly the majority). 
We may also want to look at gearbox, as spelling for manuel is ?mispelled (manual). unrepaired_damage can be explored too (nein)
There are several numeric columns that need to be converted + rename the columns.

In [57]:
autos['seller'].value_counts() #let's remove these

seller
privat        49999
gewerblich        1
Name: count, dtype: int64

In [197]:
autos.pop('seller')

0        privat
1        privat
2        privat
3        privat
4        privat
          ...  
49995    privat
49996    privat
49997    privat
49998    privat
49999    privat
Name: seller, Length: 50000, dtype: object

In [59]:
autos['offer_type'].value_counts() #let's remove these

offer_type
Angebot    49999
Gesuch         1
Name: count, dtype: int64

In [198]:
autos.pop('offer_type')

0        Angebot
1        Angebot
2        Angebot
3        Angebot
4        Angebot
          ...   
49995    Angebot
49996    Angebot
49997    Angebot
49998    Angebot
49999    Angebot
Name: offer_type, Length: 50000, dtype: object

In [62]:
autos['num_pictures'].value_counts() #this can be removed

num_pictures
0    50000
Name: count, dtype: int64

In [199]:
autos.pop('num_pictures')

0        0
1        0
2        0
3        0
4        0
        ..
49995    0
49996    0
49997    0
49998    0
49999    0
Name: num_pictures, Length: 50000, dtype: int64

In [64]:
autos['unrepaired_damage'].value_counts() #let's convert nein to no and ja to yes

unrepaired_damage
nein    35232
ja       4939
Name: count, dtype: int64

In [200]:
yes_no = {'nein':'no','ja':'yes'}
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(yes_no)

In [68]:
autos['unrepaired_damage'].value_counts()

unrepaired_damage
no     35232
yes     4939
Name: count, dtype: int64

In [72]:
autos['abtest'].value_counts() #this is ok
autos['vehicle_type'].value_counts() #this is ok
autos['registration_year'].value_counts() #this will need a closer look as there's year 1000 

registration_year
2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
        ... 
1000       1
1929       1
1927       1
1948       1
1001       1
Name: count, Length: 97, dtype: int64

In [75]:
autos['date_crawled'].value_counts()
autos['price'].value_counts() #may want to look at 0 rows

price
$0         1421
$500        781
$1,500      734
$2,500      643
$1,000      639
           ... 
$414          1
$79,933       1
$5,198        1
$18,890       1
$16,995       1
Name: count, Length: 2357, dtype: int64

In [201]:
autos['gearbox'].value_counts() #lets convert these
gear = {'manuell':'manual','automatik':'automatic'}
autos['gearbox'] = autos['gearbox'].map(gear)

In [80]:
autos['gearbox'].value_counts()

gearbox
manual       36993
automatic    10327
Name: count, dtype: int64

In [84]:
autos['power_ps'].value_counts()
autos['model'].value_counts()
autos['odometer'].value_counts()
autos['registration_month'].value_counts() #will need to look at 0 months

registration_month
0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: count, dtype: int64

In [89]:
autos['fuel_type'].value_counts()
autos['brand'].value_counts()
autos['ad_created'].value_counts()
autos['postal_code'].value_counts()
autos['last_seen'].value_counts() # the rest of these look good

last_seen
2016-04-07 06:17:27    8
2016-04-07 03:16:17    7
2016-04-06 21:17:51    7
2016-04-06 06:17:24    7
2016-04-06 13:17:03    6
                      ..
2016-04-02 11:15:28    1
2016-04-06 01:50:23    1
2016-03-21 23:45:11    1
2016-03-29 20:25:43    1
2016-04-06 21:18:48    1
Name: count, Length: 39481, dtype: int64

Summary:
price - remove $ and change to float, explore 0$
odometer - remove km and comma + rename column, convert to int
date_crawled, date_created, last_seen - these may need to be converted to datetime
explore registration_year = year 1000 noted
explore registration_month = noted 0 month
Explore null/missing data - vehicleType, gearbox, model, fuelType, notRepairedDamage

#### Price - remove $ and comma, convert to float, explore 0 dollar counts

In [202]:
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','')

In [203]:
autos['price'] = autos['price'].astype(float)

In [204]:
autos['price'].value_counts()

price
0.0        1421
500.0       781
1500.0      734
2500.0      643
1000.0      639
           ... 
414.0         1
79933.0       1
5198.0        1
18890.0       1
16995.0       1
Name: count, Length: 2357, dtype: int64

In [138]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

price
99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: count, dtype: int64

In [139]:
autos['price'].value_counts().sort_index(ascending=False).tail(20)

price
35.0       1
30.0       7
29.0       1
25.0       5
20.0       4
18.0       1
17.0       3
15.0       2
14.0       1
13.0       2
12.0       3
11.0       2
10.0       7
9.0        1
8.0        1
5.0        2
3.0        1
2.0        3
1.0      156
0.0     1421
Name: count, dtype: int64

Some of these prices can be outliers. Most expensive car in history is worth 143million, sold at auction. It is hard to imagine cars this expensive sold on ebay.

In [146]:
autos[autos['price'] >= 999999.0].head(10)

date_crawled                                               name  \
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   

            price   abtest vehicle_type  registration_year    gearbox  \
514      999999.0     test        kombi               2009     manual   
2897   11111111.0     test    limousine               1973     manual   
7814    1300000.0  control        coupe               1992        NaN   
11137  10000000.0  control        coupe               1960     manual   
22947   1234566.0  control        kombi               1999  automatic   
24384  11111111.0     test          NaN               1995        NaN   
27371  12345678.0  control          NaN               2017        NaN   
39377  12345678.0  control          NaN               2018     manual   
39705  99999999.0  control    limousine               1999  automatic   
42221  27322222.0  control    limousine               2014     manual   

       power_ps     model  odometer_km  registration_month fuel_type  \
514         101     focus       125000                   4    benzin   
2897         48    escort        50000                   3    benzin   
7814          0       NaN        50000                  12       NaN   
11137       368       NaN       100000                   1    benzin   
22947       190       NaN       150000                   2    diesel   
24384         0       NaN       150000                   0       NaN   
27371        95     punto       150000                   0       NaN   
39377        95       v40       150000                   6       NaN   
39705       224  s_klasse       150000                   9    benzin   
42221       163        c4        40000                   2    diesel   

                brand unrepaired_damage           ad_created  postal_code  \
514              ford                no  2016-03-17 00:00:00        12205   
2897             ford                no  2016-03-12 00:00:00        94469   
7814   sonstige_autos                no  2016-04-04 00:00:00        60598   
11137  sonstige_autos                no  2016-03-29 00:00:00        73033   
22947             bmw               NaN  2016-03-22 00:00:00        17454   
24384      volkswagen               NaN  2016-03-21 00:00:00        18519   
27371            fiat               NaN  2016-03-09 00:00:00        96110   
39377           volvo                no  2016-03-08 00:00:00        14542   
39705   mercedes_benz               NaN  2016-03-22 00:00:00        73525   
42221         citroen               NaN  2016-03-08 00:00:00        76532   

                 last_seen  
514    2016-04-06 07:17:35  
2897   2016-03-12 22:45:27  
7814   2016-04-05 11:34:11  
11137  2016-04-06 21:18:11  
22947  2016-04-02 03:17:32  
24384  2016-03-21 14:40:18  
27371  2016-03-09 15:45:47  
39377  2016-04-06 23:17:31  
39705  2016-04-06 05:15:30  
42221  2016-03-08 20:39:05

Looking up the car models and price - some of them appear inflated.
We will eliminate the upper end outliers and lower end outliers for sake of practice.
For the upper end, we will take 350000

In [151]:
autos['price'].value_counts().sort_index(ascending=False).tail(100)

price
329.0       2
325.0       5
320.0      12
310.0       1
300.0     384
         ... 
5.0         2
3.0         1
2.0         3
1.0       156
0.0      1421
Name: count, Length: 100, dtype: int64

For the lower end, we will take 350 (arbitrary)

In [205]:
autos = autos[(autos['price']<350000) & (autos['price']>350)]

In [155]:
autos['price'].value_counts().sort_index(ascending=False).tail(10)

price
385.0     1
380.0    29
379.0     1
378.0     1
375.0     7
370.0    21
369.0     1
360.0     8
359.0     1
356.0     1
Name: count, dtype: int64

In [206]:
autos['price'].value_counts().sort_index(ascending=False).head(100)

price
345000.0    1
299000.0    1
295000.0    1
265000.0    1
259000.0    1
           ..
64280.0     1
63999.0     2
63499.0     1
63000.0     1
62900.0     2
Name: count, Length: 100, dtype: int64

#### odometer - remove km and comma + rename column, convert to int

In [207]:
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].astype(int)
autos.rename({'odometer':'odometer_km'}, axis=1,inplace=True)

In [172]:
autos['odometer_km'].value_counts()

odometer_km
150000    29499
125000     4909
100000     2052
90000      1693
80000      1396
70000      1196
60000      1137
50000       999
40000       809
30000       769
20000       731
5000        648
10000       239
Name: count, dtype: int64

odometer column generally looks like no significant outliers

#### date_crawled, ad_created, last_seen - lets convert these to datetime

In [208]:
autos[['date_crawled','ad_created','last_seen']]

date_crawled           ad_created            last_seen
0      2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1      2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2      2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3      2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4      2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50
...                    ...                  ...                  ...
49995  2016-03-27 14:38:19  2016-03-27 00:00:00  2016-04-01 13:47:40
49996  2016-03-28 10:50:25  2016-03-28 00:00:00  2016-04-02 14:18:02
49997  2016-04-02 14:44:48  2016-04-02 00:00:00  2016-04-04 11:47:27
49998  2016-03-08 19:25:42  2016-03-08 00:00:00  2016-04-05 16:45:07
49999  2016-03-14 00:42:12  2016-03-13 00:00:00  2016-04-06 21:18:48

[46077 rows x 3 columns]

In [209]:
autos['ad_created']= pd.to_datetime(autos['ad_created'], format='%Y-%m-%d %H:%M:%S')

In [210]:
autos['date_crawled']= pd.to_datetime(autos['date_crawled'], format='%Y-%m-%d %H:%M:%S')
autos['last_seen']= pd.to_datetime(autos['last_seen'], format='%Y-%m-%d %H:%M:%S')

In [211]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 46077 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   date_crawled        46077 non-null  datetime64[ns]
 1   name                46077 non-null  object        
 2   price               46077 non-null  float64       
 3   abtest              46077 non-null  object        
 4   vehicle_type        42205 non-null  object        
 5   registration_year   46077 non-null  int64         
 6   gearbox             44166 non-null  object        
 7   power_ps            46077 non-null  int64         
 8   model               43922 non-null  object        
 9   odometer_km         46077 non-null  int32         
 10  registration_month  46077 non-null  int64         
 11  fuel_type           42688 non-null  object        
 12  brand               46077 non-null  object        
 13  unrepaired_damage   38135 non-null  object        


#### explore registration_year = year 1000 noted

In [217]:
autos['registration_year'].value_counts().sort_index().tail(20)

registration_year
2010    1587
2011    1617
2012    1307
2013     797
2014     655
2015     366
2016    1031
2017    1334
2018     466
2019       1
2800       1
4100       1
4500       1
4800       1
5000       2
5911       1
6200       1
8888       1
9000       1
9999       3
Name: count, dtype: int64

In [218]:
autos['registration_year'].value_counts().sort_index().head(20)

registration_year
1000    1
1001    1
1111    1
1800    2
1910    2
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
1950    1
1951    2
1952    1
1953    1
1954    2
Name: count, dtype: int64

we will remove anything that is in the future + after ad_created(can't be registered before ad_created) (after 2016) and before 1910 (cars werent invented until 1886)

In [222]:
autos = autos[(autos['registration_year'] >1910) & (autos['registration_year'] <2016)]

In [229]:
autos['registration_year'].describe()

count    43225.000000
mean      2002.836784
std          6.854753
min       1927.000000
25%       1999.000000
50%       2003.000000
75%       2007.000000
max       2015.000000
Name: registration_year, dtype: float64

#### explore registration_month = noted 0 month

In [230]:
autos['registration_month'].value_counts()

registration_month
3     4541
6     3891
4     3669
5     3668
7     3524
10    3305
9     3082
11    3077
12    3076
0     2971
1     2901
8     2838
2     2682
Name: count, dtype: int64

In [231]:
autos[autos['registration_month'] == 0] #0 possibly just means unknown, so we will change 0 to null to prevent confusion

date_crawled                                               name  \
46    2016-03-31 10:53:28                                       BMW_mit__Lpg   
52    2016-03-25 18:50:03  Senator_A_3.0E_Karosserie_restauriert_m._viele...   
60    2016-03-23 21:55:29                              VW_Vento_1_8_Tuev_NEU   
79    2016-03-28 19:58:42                             Seat_Ibiza_1.9_TDI__Fr   
92    2016-03-25 19:47:32                                       Zu_verkaufen   
...                   ...                                                ...   
49909 2016-04-02 20:49:31                              Honda_civic_Hatchback   
49926 2016-03-13 00:36:56                                     Vw_golf_cabrio   
49932 2016-03-14 18:52:18               Mercedes_320_cdi_7_Sitzer_Avantgarde   
49937 2016-03-20 15:47:03                        A6_TDI_2_5_l___Baujahr_1996   
49985 2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   

        price   abtest vehicle_type  registration_year    gearbox  power_ps  \
46     9000.0  control          NaN               2005  automatic         0   
52     3500.0     test    limousine               1985        NaN         0   
60     1199.0     test    limousine               1996        NaN        90   
79     6300.0  control   kleinwagen               2007     manual       131   
92     1900.0     test          NaN               2000        NaN         0   
...       ...      ...          ...                ...        ...       ...   
49909  1000.0     test       andere               1995     manual        75   
49926   850.0  control          NaN               2000        NaN         0   
49932  1400.0     test        kombi               1999  automatic         0   
49937   999.0  control        kombi               1996     manual       136   
49985  1000.0  control          NaN               1995  automatic         0   

        model  odometer_km  registration_month fuel_type          brand  \
46        NaN       150000                   0       NaN            bmw   
52     andere         5000                   0    benzin           opel   
60        NaN       150000                   0    benzin     volkswagen   
79      ibiza       150000                   0    diesel           seat   
92        NaN       150000                   0       NaN            bmw   
...       ...          ...                 ...       ...            ...   
49909   civic       150000                   0    benzin          honda   
49926     NaN       150000                   0       NaN     volkswagen   
49932     NaN       150000                   0       NaN  mercedes_benz   
49937      a6       150000                   0    diesel           audi   
49985     NaN       150000                   0    benzin     volkswagen   

      unrepaired_damage ad_created  postal_code           last_seen  
46                  NaN 2016-03-31        12351 2016-04-06 03:44:41  
52                   no 2016-03-25        63500 2016-04-07 00:46:00  
60                   no 2016-03-23         1665 2016-04-06 05:45:36  
79                  NaN 2016-03-28         4157 2016-04-07 02:17:59  
92                  NaN 2016-03-25        72108 2016-04-07 02:16:01  
...                 ...        ...          ...                 ...  
49909                no 2016-04-02        15517 2016-04-06 21:44:23  
49926               NaN 2016-03-12        53227 2016-03-15 17:17:49  
49932               NaN 2016-03-14        94315 2016-03-14 18:52:18  
49937               yes 2016-03-20        83539 2016-04-06 20:18:35  
49985               NaN 2016-04-02        30900 2016-04-06 15:17:52  

[2971 rows x 17 columns]

In [233]:
autos.loc[autos['registration_month'] == 0, 'registration_month'] = None

In [236]:
autos['registration_month'].value_counts(dropna=False)

registration_month
3.0     4541
6.0     3891
4.0     3669
5.0     3668
7.0     3524
10.0    3305
9.0     3082
11.0    3077
12.0    3076
NaN     2971
1.0     2901
8.0     2838
2.0     2682
Name: count, dtype: int64

#### Explore null/missing data - vehicle_type, gearbox, model, fuel_type, unrepaired_damage

In [238]:
autos['vehicle_type'].value_counts(dropna=False)

vehicle_type
limousine     12233
kleinwagen     9584
kombi          8678
bus            3975
cabrio         2982
coupe          2395
suv            1952
NaN            1067
andere          359
Name: count, dtype: int64

In [239]:
autos['gearbox'].value_counts(dropna=False)

gearbox
manual       32079
automatic     9582
NaN           1564
Name: count, dtype: int64

In [240]:
autos['model'].value_counts(dropna=False)

model
golf          3413
andere        3163
3er           2514
NaN           1780
polo          1364
              ... 
kappa            2
charade          2
rangerover       1
200              1
b_max            1
Name: count, Length: 244, dtype: int64

In [241]:
autos['fuel_type'].value_counts(dropna=False)

fuel_type
benzin     26374
diesel     13797
NaN         2295
lpg          629
cng           67
hybrid        36
elektro       16
andere        11
Name: count, dtype: int64

In [242]:
autos['unrepaired_damage'].value_counts(dropna=False)

unrepaired_damage
no     32813
NaN     6719
yes     3693
Name: count, dtype: int64

null values likely to be unknown data, unless it affects further analysis, we will keep them for now

### Data Analysis <a class='anchor' id='4'></a>

Now that the dataset is finally clean, let's proceed with data analysis!

#### Exploring Price by Brand

In [246]:
autos['brand'].info()

<class 'pandas.core.series.Series'>
Index: 43225 entries, 0 to 49999
Series name: brand
Non-Null Count  Dtype 
--------------  ----- 
43225 non-null  object
dtypes: object(1)
memory usage: 675.4+ KB


In [268]:
top_20_index = autos['brand'].value_counts().head(20).index #let's look at the top 20 brands
top_20_index = list(top_20_index)
top_20_index

['volkswagen',
 'bmw',
 'opel',
 'mercedes_benz',
 'audi',
 'ford',
 'renault',
 'peugeot',
 'fiat',
 'seat',
 'skoda',
 'mazda',
 'nissan',
 'smart',
 'citroen',
 'toyota',
 'hyundai',
 'sonstige_autos',
 'volvo',
 'mini']

#### Calculating avg_price / brand

In [273]:
brand_bool = autos['brand'].isin(top_20_index)
autos_brand = autos.loc[brand_bool]

In [280]:
mean_autos_by_brand = autos_brand.groupby(by='brand')['price'].mean().sort_values(ascending=False)

In [285]:
autos_brand.groupby(by='brand')['price'].describe().sort_values(by='mean', ascending=False)

count          mean           std    min      25%     50%  \
brand                                                                        
sonstige_autos   417.0  12686.465228  21874.852955  370.0  2999.00  7800.0   
mini             397.0  10706.408060   5931.129801  900.0  6222.00  9250.0   
audi            3898.0   9509.454336   9668.529969  380.0  2500.00  6500.0   
mercedes_benz   4377.0   8805.609321  10227.115261  370.0  2499.00  5400.0   
bmw             4952.0   8567.467488   9915.179969  360.0  2500.00  6000.0   
skoda            735.0   6525.182313   5184.458799  400.0  2750.00  5000.0   
volkswagen      9120.0   5740.838925   6269.148250  356.0  1499.00  3350.0   
hyundai          443.0   5632.036117   4803.866079  390.0  1999.00  4500.0   
toyota           572.0   5300.666084   4724.996619  370.0  2075.00  4300.0   
volvo            405.0   5172.656790   5591.651494  399.0  1299.00  2990.0   
nissan           651.0   5117.929339   5604.679469  380.0  1100.00  2750.0   
seat             773.0   4695.868047   4713.640919  380.0  1250.00  3199.0   
mazda            653.0   4403.356815   4992.319263  399.0  1200.00  2800.0   
ford            2856.0   4211.240196   6041.391457  360.0  1000.00  2200.0   
citroen          610.0   3987.514754   3869.483552  399.0  1407.50  2850.0   
smart            639.0   3638.244131   2308.582583  400.0  1869.00  2998.0   
peugeot         1262.0   3323.768621   3158.564137  370.0  1199.25  2350.0   
opel            4382.0   3298.394112   3665.340551  359.0   990.00  1990.0   
fiat            1031.0   3172.506305   3088.362214  370.0   999.00  2000.0   
renault         1929.0   2712.309487   3655.562868  360.0   900.00  1500.0   

                    75%       max  
brand                              
sonstige_autos  14500.0  345000.0  
mini            14250.0   34500.0  
audi            13175.0   80000.0  
mercedes_benz   11950.0  180000.0  
bmw             11500.0  259000.0  
skoda            8500.0   32700.0  
volkswagen       7950.0   64500.0  
hyundai          7675.0   29999.0  
toyota           6599.0   49500.0  
volvo            6900.0   29500.0  
nissan           7325.0   34999.0  
seat             6500.0   25900.0  
mazda            5500.0   38900.0  
ford             4990.0  130000.0  
citroen          5000.0   33000.0  
smart            5000.0   16850.0  
peugeot          4300.0   25500.0  
opel             4200.0   38990.0  
fiat             4200.0   25000.0  
renault          3300.0   93000.0

It appears among the top 20, sonstige_autos, mini, audi, mercedes, bmw are in the top 5 in terms of mean price. Also note that the std are quite large for these data. For a lot of the brand it looks like the median 50% is smaller than the mean, this possibly suggest a right skew. There may be a few large outliers producing these results and also increasing the std dev.

#### Next we look at avg mileage (odometer_km) and compare it to average price for the top 5 brands

In [324]:
top_5_autos_brand = list(mean_autos_by_brand.iloc[:5].index)
top_5_autos_brand

['sonstige_autos', 'mini', 'audi', 'mercedes_benz', 'bmw']

In [325]:
top_5_autos_brand_data = autos.loc[autos['brand'].isin(top_5_autos_brand)]

In [326]:
top_5_autos_brand_data_sorted = top_5_autos_brand_data.groupby('brand')[['price','odometer_km']].mean()
top_5_autos_brand_data_sorted.rename({'price':'avg_price','odometer_km':'avg_mileage'}, axis=1, inplace=True)

In [327]:
top_5_autos_brand_data_sorted

avg_price    avg_mileage
brand                                      
audi             9509.454336  129129.040534
bmw              8567.467488  132803.917609
mercedes_benz    8805.609321  130924.148960
mini            10706.408060   88816.120907
sonstige_autos  12686.465228   91223.021583

In [328]:
top_5_autos_brand_data_sorted['avg_price'] = top_5_autos_brand_data_sorted['avg_price'].apply('{:.2f}'.format)
top_5_autos_brand_data_sorted['avg_mileage'] = top_5_autos_brand_data_sorted['avg_mileage'].apply('{:.2f}'.format)

In [329]:
top_5_autos_brand_data_sorted.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5 entries, audi to sonstige_autos
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   avg_price    5 non-null      object
 1   avg_mileage  5 non-null      object
dtypes: object(2)
memory usage: 120.0+ bytes


In [332]:
top_5_autos_brand_data_sorted = top_5_autos_brand_data_sorted.astype(float)

In [333]:
top_5_autos_brand_data_sorted.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5 entries, audi to sonstige_autos
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   avg_price    5 non-null      float64
 1   avg_mileage  5 non-null      float64
dtypes: float64(2)
memory usage: 120.0+ bytes


In [334]:
top_5_autos_brand_data_sorted

avg_price  avg_mileage
brand                                 
audi              9509.45    129129.04
bmw               8567.47    132803.92
mercedes_benz     8805.61    130924.15
mini             10706.41     88816.12
sonstige_autos   12686.47     91223.02

There is an indirect relationship between avg_price and avg_mileage. As mileages goes up, price goes down (not unexpected).

#### Conclusion <a class='anchor' id='5'></a>

In this project, we have gone through the process of data cleaning and performed simple data analysis on the dataset at the end.

Specifically, we practiced:
- Renaming columns
- Removing columns
- Renaming elements
- Modifying and standardizing str elements
- Removing outliers
- Datetime conversion